File: lld/problems/kafka-lld/java/src/Topic.java
import java.util.ArrayList;
import java.util.List;

/** Owns a fixed set of partitions created at topic-creation time. */
public class Topic {
    public static class InvalidPartitionCountException extends RuntimeException {
        public InvalidPartitionCountException() {
            super("numPartitions must be > 0");
        }
    }

    private final String name;
    private final List<Partition> partitions;
    private final PartitionerStrategy partitioner;

    public Topic(String name, int numPartitions, PartitionerStrategy partitioner) {
        if (numPartitions <= 0) {
            throw new InvalidPartitionCountException();
        }
        this.name = name;
        this.partitioner = partitioner;
        this.partitions = new ArrayList<>(numPartitions);
        for (int i = 0; i < numPartitions; i++) {
            this.partitions.add(new Partition());
        }
    }

    public String getName() {
        return name;
    }

    public List<Partition> getPartitions() {
        return partitions;
    }

    public int selectPartition(String key) {
        return partitioner.partition(key, partitions.size());
    }
}

File: lld/problems/kafka-lld/go/topic.go
package kafkalld

import "errors"

var ErrInvalidPartitionCount = errors.New("numPartitions must be > 0")

// Topic owns a fixed set of partitions created at topic-creation time.
type Topic struct {
	Name        string
	Partitions  []*Partition
	partitioner PartitionerStrategy
}

func NewTopic(name string, numPartitions int, partitioner PartitionerStrategy) (*Topic, error) {
	if numPartitions <= 0 {
		return nil, ErrInvalidPartitionCount
	}
	partitions := make([]*Partition, numPartitions)
	for i := range partitions {
		partitions[i] = NewPartition()
	}
	return &Topic{Name: name, Partitions: partitions, partitioner: partitioner}, nil
}

func (t *Topic) SelectPartition(key string) int {
	return t.partitioner.Partition(key, len(t.Partitions))
}
